src/System.Private.CoreLib/shared/System/Threading/Volatile.cs

```diff
@@ -0,0 +1,198 @@
+        #endregion
+
+        #region Byte
+        private struct VolatileByte { public volatile byte Value; }
+
+        [NonVersionable]
+        public static byte Read(ref byte location) =>
+            Unsafe.As<byte, VolatileByte>(ref location).Value;
+
+        [NonVersionable]
+        public static void Write(ref byte location, byte value) =>
+            Unsafe.As<byte, VolatileByte>(ref location).Value = value;
+        #endregion
+
+        #region Double
+        [NonVersionable]
+        public static double Read(ref double location)
+        {
+            long result = Read(ref Unsafe.As<double, long>(ref location));
+            return *(double*)&result;
+        }
+
+        [NonVersionable]
+        public static void Write(ref double location, double value) =>
+            Write(ref Unsafe.As<double, long>(ref location), *(long*)&value);
+        #endregion
+
+        #region Int16
+        private struct VolatileInt16 { public volatile short Value; }
+
+        [NonVersionable]
+        public static short Read(ref short location) =>
+            Unsafe.As<short, VolatileInt16>(ref location).Value;
+
+        [NonVersionable]
+        public static void Write(ref short location, short value) =>
+            Unsafe.As<short, VolatileInt16>(ref location).Value = value;
+        #endregion
+
+        #region Int32
+        private struct VolatileInt32 { public volatile int Value; }
+
+        [NonVersionable]
+        public static int Read(ref int location) =>
+            Unsafe.As<int, VolatileInt32>(ref location).Value;
+
+        [NonVersionable]
+        public static void Write(ref int location, int value) =>
+            Unsafe.As<int, VolatileInt32>(ref location).Value = value;
+        #endregion
+
+        #region Int64
+        [NonVersionable]
+        public static long Read(ref long location) =>
+#if BIT64
+            (Int64)Unsafe.As<Int64, VolatileIntPtr>(ref location).Value;
+#else
+            // On 32-bit machines, we use Interlocked, since an ordinary volatile read would not be atomic.
+            Interlocked.CompareExchange(ref location, 0, 0);
+#endif
+
+        [NonVersionable]
+        public static void Write(ref long location, long value) =>
+#if BIT64
+            Unsafe.As<Int64, VolatileIntPtr>(ref location).Value = (IntPtr)value;
+#else
+            // On 32-bit, we use Interlocked, since an ordinary volatile write would not be atomic.
+            Interlocked.Exchange(ref location, value);
+#endif
+        #endregion
+
+        #region IntPtr
+        private struct VolatileIntPtr { public volatile IntPtr Value; }
+
+        [NonVersionable]
+        public static IntPtr Read(ref IntPtr location) =>
+            Unsafe.As<IntPtr, VolatileIntPtr>(ref location).Value;
+
+        [NonVersionable]
+        public static void Write(ref IntPtr location, IntPtr value) =>
+            Unsafe.As<IntPtr, VolatileIntPtr>(ref location).Value = value;
```

Comment:
@jkotas, do you know why corert's implementation has this?
https://github.com/dotnet/corert/blob/382f0fe4548f6bc9c1bc150f3391075527af59fe/src/System.Private.CoreLib/src/System/Threading/Volatile.cs#L118-L121
I'm not sure what's special about this one case that required using fixed like that.

Reply:
It looks like an oversight.